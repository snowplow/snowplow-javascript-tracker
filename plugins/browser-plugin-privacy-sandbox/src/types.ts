export type PrivacySandboxTopic = {
  configVersion: string;
  modelVersion: string;
  taxonomyVersion: string;
  topic: number;
  version: string;
};

export type Topic = Pick<PrivacySandboxTopic, 'topic' | 'version'>;

export type BrowsingTopicsOptions = {
  /* Enabling will not cause the current page to be included in the weekly epoch calculation or update the list of topics observed for the caller. */
  skipObservation?: boolean;
};
