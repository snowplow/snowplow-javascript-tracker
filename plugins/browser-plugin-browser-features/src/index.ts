/*
 * Copyright (c) 2022 Snowplow Analytics Ltd, 2010 Anthon Pang
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are met:
 *
 * 1. Redistributions of source code must retain the above copyright notice, this
 *    list of conditions and the following disclaimer.
 *
 * 2. Redistributions in binary form must reproduce the above copyright notice,
 *    this list of conditions and the following disclaimer in the documentation
 *    and/or other materials provided with the distribution.
 *
 * 3. Neither the name of the copyright holder nor the names of its
 *    contributors may be used to endorse or promote products derived from
 *    this software without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
 * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
 * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
 * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
 * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
 * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
 * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

import { isFunction, BrowserPlugin, BrowserTracker } from '@snowplow/browser-tracker-core';

declare global {
  interface MimeTypeArray {
    [index: string]: MimeType;
  }
}

const pluginMap: Record<string, string> = {
  // document types
  pdf: 'application/pdf',

  // media players
  qt: 'video/quicktime',
  realp: 'audio/x-pn-realaudio-plugin',
  wma: 'application/x-mplayer2',

  // interactive multimedia
  dir: 'application/x-director',
  fla: 'application/x-shockwave-flash',

  // RIA
  java: 'application/x-java-vm',
  gears: 'application/x-googlegears',
  ag: 'application/x-silverlight',
};

/**
 * Adds the available MIME Types to each event
 */
export function BrowserFeaturesPlugin(): BrowserPlugin {
  return {
    activateBrowserPlugin: (tracker: BrowserTracker) => {
      const navigatorAlias = navigator;
      // General plugin detection
      if (navigatorAlias.mimeTypes && navigatorAlias.mimeTypes.length) {
        for (const i in pluginMap) {
          if (Object.prototype.hasOwnProperty.call(pluginMap, i)) {
            let mimeType = navigatorAlias.mimeTypes[pluginMap[i]];
            if (mimeType) {
              tracker.core.addPayloadPair('f_' + i, mimeType.enabledPlugin ? '1' : '0');
            }
          }
        }
      }

      // Firefox
      if (isFunction((window as any).GearsFactory)) {
        tracker.core.addPayloadPair('f_gears', '1');
      }
    },
  };
}
