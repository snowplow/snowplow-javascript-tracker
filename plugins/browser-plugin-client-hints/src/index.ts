import { BrowserPlugin } from '@snowplow/browser-tracker-core';
import { HttpClientHints } from './contexts';

interface NavigatorUABrandVersion {
  brand: string; // "Google Chrome"
  version: string; // "84"
}

interface UADataValues {
  platform: string; // "PhoneOS"
  platformVersion: string; // "10A"
  architecture: string; // "arm"
  model: string; // "X644GTM"
  uaFullVersion: string; // "73.32.AGX.5"
}

declare global {
  interface Navigator {
    userAgentData: {
      mobile: boolean;
      brands: Array<NavigatorUABrandVersion>;
      getHighEntropyValues: (hints: Array<string>) => Promise<UADataValues>;
    };
  }
}

let uaClientHints: HttpClientHints;

/**
 * This function makes sure that the expected array is returned as an array instead of an object.
 * It handles a problem that in some cases the `navigator.userAgentData.brands` was returned as an object instead of array.
 */
function forceArray<T>(array: T[] | Record<string, T>): T[] {
  if (Array.isArray(array)) return array;

  if (Object.prototype.toString.call(array) === '[object Object]') {
    return Object.keys(array).map((e) => {
      return array[e];
    });
  }

  return [];
}

/**
 * Returns the client-hints brands, ensuring no additional properties.
 */
function getBrands(brands: Array<NavigatorUABrandVersion>): Array<NavigatorUABrandVersion> {
  return brands.map((b) => {
    const { brand, version } = b;
    return { brand, version };
  });
}

/**
 * Validates whether userAgentData is compliant to the client-hints interface.
 * https://wicg.github.io/ua-client-hints/#interface
 */
function validClientHints(hints: HttpClientHints): boolean {
  if (!hints || typeof hints.isMobile !== 'boolean' || !Array.isArray(hints.brands)) {
    return false;
  }

  if (
    hints.brands.length === 0 ||
    hints.brands.some((brand) => typeof brand.brand !== 'string' || typeof brand.version !== 'string')
  ) {
    return false;
  }

  return true;
}

/**
 * Attaches Client Hint information where available
 * @param includeHighEntropy - Should high entropy values be included
 */
export function ClientHintsPlugin(includeHighEntropy?: boolean): BrowserPlugin {
  const populateClientHints = () => {
    const navigatorAlias = navigator;
    const uaData = navigatorAlias.userAgentData;

    if (uaData) {
      let candidateHints: HttpClientHints;
      candidateHints = {
        isMobile: uaData.mobile,
        brands: getBrands(forceArray(uaData.brands)),
      };
      if (includeHighEntropy && uaData.getHighEntropyValues) {
        uaData
          .getHighEntropyValues(['platform', 'platformVersion', 'architecture', 'model', 'uaFullVersion'])
          .then((res) => {
            candidateHints.architecture = res.architecture;
            candidateHints.model = res.model;
            candidateHints.platform = res.platform;
            candidateHints.uaFullVersion = res.uaFullVersion;
            candidateHints.platformVersion = res.platformVersion;
          });
      }

      if (validClientHints(candidateHints)) {
        uaClientHints = candidateHints;
      }
    }
  };

  return {
    activateBrowserPlugin: () => {
      if (!uaClientHints) {
        populateClientHints();
      }
    },
    contexts: () => {
      if (uaClientHints) {
        return [
          {
            schema: 'iglu:org.ietf/http_client_hints/jsonschema/1-0-0',
            data: uaClientHints,
          },
        ];
      }
      return [];
    },
  };
}
