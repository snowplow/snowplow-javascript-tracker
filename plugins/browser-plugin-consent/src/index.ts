/*
 * Copyright (c) 2022 Snowplow Analytics Ltd, 2010 Anthon Pang
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are met:
 *
 * 1. Redistributions of source code must retain the above copyright notice, this
 *    list of conditions and the following disclaimer.
 *
 * 2. Redistributions in binary form must reproduce the above copyright notice,
 *    this list of conditions and the following disclaimer in the documentation
 *    and/or other materials provided with the distribution.
 *
 * 3. Neither the name of the copyright holder nor the names of its
 *    contributors may be used to endorse or promote products derived from
 *    this software without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
 * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
 * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
 * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
 * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
 * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
 * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

import { BrowserPlugin, BrowserTracker, dispatchToTrackersInCollection } from '@snowplow/browser-tracker-core';
import {
  buildConsentGranted,
  buildConsentWithdrawn,
  CommonEventProperties,
  ConsentGrantedEvent,
  ConsentWithdrawnEvent,
  Logger,
} from '@snowplow/tracker-core';
import { Gdpr } from './contexts';

export { ConsentGrantedEvent, ConsentWithdrawnEvent };

export enum gdprBasis {
  consent = 'consent',
  contract = 'contract',
  legalObligation = 'legal_obligation',
  vitalInterests = 'vital_interests',
  publicTask = 'public_task',
  legitimateInterests = 'legitimate_interests',
}

export type GdprBasis = keyof typeof gdprBasis;

/** The Configuration for the GDPR Context */
export interface GdprContextConfiguration {
  /** The basis for why the document will be processed */
  basisForProcessing: GdprBasis;
  /** An identifier for the document */
  documentId?: string;
  /** The version of the document */
  documentVersion?: string;
  /** A descrtiption of the document */
  documentDescription?: string;
}

const _trackers: Record<string, BrowserTracker> = {};
const _context: Record<string, Gdpr> = {};
let LOG: Logger;

/**
 * The Consent Plugin
 *
 * Adds Consent Granted and Withdrawn events
 * and the ability to add the GDPR context to events
 */
export function ConsentPlugin(): BrowserPlugin {
  let trackerId: string;

  return {
    activateBrowserPlugin: (tracker) => {
      trackerId = tracker.id;
      _trackers[tracker.id] = tracker;
    },
    contexts: () => {
      if (_context[trackerId]) {
        return [
          {
            schema: 'iglu:com.snowplowanalytics.snowplow/gdpr/jsonschema/1-0-0',
            data: _context[trackerId],
          },
        ];
      }

      return [];
    },
    logger: (logger) => {
      LOG = logger;
    },
  };
}

/**
 * Enable the GDPR context for each event
 * @param configuration - the configuration for the GDPR context
 * @param trackers - The tracker identifiers which should have the GDPR context enabled
 */
export function enableGdprContext(
  configuration: GdprContextConfiguration,
  trackers: Array<string> = Object.keys(_trackers)
) {
  const { basisForProcessing, documentId, documentVersion, documentDescription } = configuration;
  let basis = gdprBasis[basisForProcessing];

  if (!basis) {
    LOG.warn(
      'enableGdprContext: basisForProcessing must be one of: consent, contract, legalObligation, vitalInterests, publicTask, legitimateInterests'
    );
    return;
  } else {
    trackers.forEach((t) => {
      if (_trackers[t]) {
        _context[t] = {
          basisForProcessing: basis,
          documentId: documentId ?? null,
          documentVersion: documentVersion ?? null,
          documentDescription: documentDescription ?? null,
        };
      }
    });
  }
}

/**
 * Track a consent granted action
 *
 * @param event - The event information
 * @param trackers - The tracker identifiers which the event will be sent to
 */
export function trackConsentGranted(
  event: ConsentGrantedEvent & CommonEventProperties,
  trackers: Array<string> = Object.keys(_trackers)
) {
  dispatchToTrackersInCollection(trackers, _trackers, (t) => {
    const builtEvent = buildConsentGranted(event);
    t.core.track(
      builtEvent.event,
      event.context ? event.context.concat(builtEvent.context) : builtEvent.context,
      event.timestamp
    );
  });
}

/**
 * Track a consent withdrawn action
 *
 * @param event - The event information
 * @param trackers - The tracker identifiers which the event will be sent to
 */
export function trackConsentWithdrawn(
  event: ConsentWithdrawnEvent & CommonEventProperties,
  trackers: Array<string> = Object.keys(_trackers)
) {
  dispatchToTrackersInCollection(trackers, _trackers, (t) => {
    const builtEvent = buildConsentWithdrawn(event);
    t.core.track(
      builtEvent.event,
      event.context ? event.context.concat(builtEvent.context) : builtEvent.context,
      event.timestamp
    );
  });
}
