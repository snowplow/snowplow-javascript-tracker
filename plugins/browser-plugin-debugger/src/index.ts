/*
 * Copyright (c) 2022 Snowplow Analytics Ltd, 2010 Anthon Pang
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are met:
 *
 * 1. Redistributions of source code must retain the above copyright notice, this
 *    list of conditions and the following disclaimer.
 *
 * 2. Redistributions in binary form must reproduce the above copyright notice,
 *    this list of conditions and the following disclaimer in the documentation
 *    and/or other materials provided with the distribution.
 *
 * 3. Neither the name of the copyright holder nor the names of its
 *    contributors may be used to endorse or promote products derived from
 *    this software without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
 * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
 * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
 * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
 * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
 * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
 * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

import { BrowserPlugin, BrowserTracker } from '@snowplow/browser-tracker-core';
import {
  EventJson,
  EventJsonWithKeys,
  JsonProcessor,
  Logger,
  LOG_LEVEL,
  Payload,
  PayloadBuilder,
  payloadJsonProcessor,
  version,
  getSchemaParts,
  SelfDescribingJson,
} from '@snowplow/tracker-core';
import randomcolor from 'randomcolor';

/**
 * Adds advertisement tracking functions
 */
export function DebuggerPlugin(logLevel: LOG_LEVEL = LOG_LEVEL.debug): BrowserPlugin {
  let LOG: Logger;
  let tracker: BrowserTracker;
  let eventColour: string;

  const colours = {
    event: () => `color: White; background: ${eventColour}; font-weight: bold; padding: 1px 4px; border-radius: 2px;`,
    snowplowPurple: 'color: White; background: #6638B8; font-weight: bold; padding: 1px 4px; border-radius: 2px;',
    activated: 'color: White; background: #9E62DD; font-weight: bold; padding: 1px 4px; border-radius: 2px;',
    payload: 'color: White; background: #3748B8; font-weight: bold; padding: 1px 4px; border-radius: 2px;',
    json: 'color: White; background: #388AB8; font-weight: bold; padding: 1px 4px; border-radius: 2px;',
    schema: 'color: White; background: #268047; font-weight: bold; padding: 1px 4px; border-radius: 2px;',
    schemaVersion: 'color: White; background: #80265F; font-weight: bold; padding: 1px 4px; border-radius: 2px;',
  };

  const debug = (style: string, message: string, extra?: [string, ...unknown[]]) => {
    const [extraMessage, ...rest] = extra ?? [''];
    LOG.debug(
      `v${version} %c${tracker.namespace}%c %c%s` + extraMessage,
      colours.snowplowPurple,
      '',
      style,
      message,
      ...rest
    );
  };

  function jsonInterceptor(encodeBase64: boolean): JsonProcessor {
    const log = (json: EventJsonWithKeys, data: SelfDescribingJson) => {
      const schemaParts = getSchemaParts(data['schema']);
      debug(colours.event(), 'Event', [
        '%c%s%c%s%c%s\n%o',
        colours.json,
        `${getJsonType(json)}: ${schemaParts ? schemaParts[1] : 'Unknown Schema'}`,
        colours.schemaVersion,
        schemaParts ? `Version: ${schemaParts[2]}-${schemaParts[3]}-${schemaParts[4]}` : 'Unknown Schema Version',
        colours.schema,
        schemaParts ? `Vendor: ${schemaParts[0]}` : 'Unknown Vendor',
        data['data'],
      ]);
    };

    return (payloadBuilder: PayloadBuilder, jsonForProcessing: EventJson) => {
      if (jsonForProcessing.length) {
        for (const json of jsonForProcessing) {
          const data = json.json['data'] as SelfDescribingJson;
          if (Array.isArray(data)) {
            data.forEach((d) => {
              log(json, d);
            });
          } else {
            log(json, data);
          }
        }
      }
      return payloadJsonProcessor(encodeBase64)(payloadBuilder, jsonForProcessing);
    };
  }

  return {
    logger: (logger) => {
      LOG = logger;
      logger.setLogLevel(logLevel);
    },
    activateBrowserPlugin: (t) => {
      tracker = t;
      debug(colours.activated, 'Tracker Activated');
    },
    beforeTrack: (payloadBuilder: PayloadBuilder) => {
      eventColour = randomcolor({ luminosity: 'dark' });
      payloadBuilder.withJsonProcessor(jsonInterceptor(tracker.core.getBase64Encoding()));
      debug(colours.event(), 'Event', ['%c%s', colours.snowplowPurple, getEventType(payloadBuilder)]);
      payloadBuilder.build();
    },
    afterTrack: (payload: Payload) => {
      debug(colours.event(), 'Event', ['%c%s\n%o', colours.payload, 'Payload', payload]);
    },
  };
}

function getJsonType(json: EventJsonWithKeys) {
  switch (json.keyIfEncoded) {
    case 'cx':
      return 'Context';
    case 'ue_px':
      return 'Self Describing';
    default:
      return `${json.keyIfEncoded}, ${json.keyIfNotEncoded}`;
  }
}

function getEventType(payloadBuilder: PayloadBuilder) {
  const payload = payloadBuilder.getPayload();
  switch (payload['e']) {
    case 'pv':
      return 'Page View';
    case 'pp':
      return 'Page Ping';
    case 'tr':
      return 'Ecommerce Transaction';
    case 'ti':
      return 'Ecommerce Transaction Item';
    case 'se':
      return 'Structured Event';
    case 'ue':
      return 'Self Describing';
    default:
      return typeof payload['e'] === 'string' ? payload['e'] : 'Invalid';
  }
}
