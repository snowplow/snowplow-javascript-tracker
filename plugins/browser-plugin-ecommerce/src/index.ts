/*
 * Copyright (c) 2022 Snowplow Analytics Ltd, 2010 Anthon Pang
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are met:
 *
 * 1. Redistributions of source code must retain the above copyright notice, this
 *    list of conditions and the following disclaimer.
 *
 * 2. Redistributions in binary form must reproduce the above copyright notice,
 *    this list of conditions and the following disclaimer in the documentation
 *    and/or other materials provided with the distribution.
 *
 * 3. Neither the name of the copyright holder nor the names of its
 *    contributors may be used to endorse or promote products derived from
 *    this software without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
 * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
 * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
 * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
 * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
 * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
 * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

import { BrowserPlugin, BrowserTracker, dispatchToTrackersInCollection } from '@snowplow/browser-tracker-core';
import {
  AddToCartEvent,
  RemoveFromCartEvent,
  buildAddToCart,
  buildEcommerceTransaction,
  buildEcommerceTransactionItem,
  buildRemoveFromCart,
  CommonEventProperties,
  EcommerceTransactionEvent,
  EcommerceTransactionItemEvent,
} from '@snowplow/tracker-core';

export { AddToCartEvent, RemoveFromCartEvent, EcommerceTransactionEvent, EcommerceTransactionItemEvent };

interface Transaction {
  transaction?: EcommerceTransactionEvent & CommonEventProperties;
  items: Array<EcommerceTransactionItemEvent & CommonEventProperties>;
}

function ecommerceTransactionTemplate(): Transaction {
  return {
    items: [],
  };
}

const _trackers: Record<string, BrowserTracker> = {};
const _transactions: Record<string, Transaction> = {};

/**
 * Adds ecommerce and cart tracking
 */
export function EcommercePlugin(): BrowserPlugin {
  return {
    activateBrowserPlugin: (tracker) => {
      _trackers[tracker.id] = tracker;
      _transactions[tracker.id] = ecommerceTransactionTemplate();
    },
  };
}

/**
 * Track an ecommerce transaction
 *
 * @param event - The event information
 * @param trackers - The tracker identifiers which the event will be sent to
 */
export function addTrans(
  event: EcommerceTransactionEvent & CommonEventProperties,
  trackers: Array<string> = Object.keys(_trackers)
) {
  trackers.forEach((t) => {
    if (_transactions[t]) {
      _transactions[t].transaction = event;
    }
  });
}

/**
 * Track an ecommerce transaction item
 *
 * @param event - The event information
 * @param trackers - The tracker identifiers which the event will be sent to
 */
export function addItem(
  event: EcommerceTransactionItemEvent & CommonEventProperties,
  trackers: Array<string> = Object.keys(_trackers)
) {
  trackers.forEach((t) => {
    if (_transactions[t]) {
      _transactions[t].items.push(event);
    }
  });
}

/**
 * Commit the ecommerce transaction
 *
 * @remarks
 * This call will send the data specified with addTrans, ddItem methods to the tracking server.
 */
export function trackTrans(trackers: Array<string> = Object.keys(_trackers)) {
  dispatchToTrackersInCollection(trackers, _trackers, (t) => {
    const transaction = _transactions[t.id].transaction;
    if (transaction) {
      t.core.track(buildEcommerceTransaction(transaction), transaction.context, transaction.timestamp);
    }
    for (var i = 0; i < _transactions[t.id].items.length; i++) {
      const item = _transactions[t.id].items[i];
      t.core.track(buildEcommerceTransactionItem(item), item.context, item.timestamp);
    }

    _transactions[t.id] = ecommerceTransactionTemplate();
  });
}

/**
 * Track an add-to-cart event
 *
 * @param event - The event information
 * @param trackers - The tracker identifiers which the event will be sent to
 */
export function trackAddToCart(
  event: AddToCartEvent & CommonEventProperties,
  trackers: Array<string> = Object.keys(_trackers)
) {
  dispatchToTrackersInCollection(trackers, _trackers, (t) => {
    t.core.track(buildAddToCart(event), event.context, event.timestamp);
  });
}

/**
 * Track a remove-from-cart event
 *
 * @param event - The event information
 * @param trackers - The tracker identifiers which the event will be sent to
 */
export function trackRemoveFromCart(
  event: RemoveFromCartEvent & CommonEventProperties,
  trackers: Array<string> = Object.keys(_trackers)
) {
  dispatchToTrackersInCollection(trackers, _trackers, (t) => {
    t.core.track(buildRemoveFromCart(event), event.context, event.timestamp);
  });
}
