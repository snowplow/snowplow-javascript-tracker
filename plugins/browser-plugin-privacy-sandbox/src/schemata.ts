export const PRIVACY_SANDBOX_TOPICS_SCHEMA = 'iglu:com.google.privacysandbox/topics/jsonschema/1-0-0';
