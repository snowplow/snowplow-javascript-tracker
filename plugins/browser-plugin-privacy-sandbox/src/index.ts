import { BrowserPlugin } from '@snowplow/browser-tracker-core';
import { PRIVACY_SANDBOX_TOPICS_SCHEMA } from './schemata';
import { LOG } from '@snowplow/tracker-core';
import { BrowsingTopicsOptions, PrivacySandboxTopic, Topic } from './types';

declare global {
  interface Document {
    /* Declare Feature Policy types https://developer.mozilla.org/en-US/docs/Web/API/FeaturePolicy */
    featurePolicy?: {
      allowsFeature: (feature: string, origin?: string) => boolean;
    };
    browsingTopics: (options: BrowsingTopicsOptions) => Promise<PrivacySandboxTopic[]>;
  }
}

/**
 * Adds Privacy Sandbox Topics context to events
 */
export function PrivacySandboxPlugin(): BrowserPlugin {
  let topics: Topic[] = [];

  return {
    activateBrowserPlugin: function () {
      const isTopicsApiAvailable =
        'browsingTopics' in document && document.featurePolicy?.allowsFeature('browsing-topics');

      if (isTopicsApiAvailable) {
        document
          .browsingTopics({ skipObservation: true })
          .then((availableTopics) => {
            topics = availableTopics.map(({ topic, version }) => ({
              topic,
              version,
            }));
          })
          .catch((err) => LOG.error(err));
      }
    },
    contexts: function () {
      return topics.length ? [{ schema: PRIVACY_SANDBOX_TOPICS_SCHEMA, data: { topics } }] : [];
    },
  };
}
