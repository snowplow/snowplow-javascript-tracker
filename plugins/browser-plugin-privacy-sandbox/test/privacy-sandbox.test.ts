import { buildLinkClick, trackerCore } from '@snowplow/tracker-core';
import { JSDOM } from 'jsdom';
import { PrivacySandboxPlugin } from '../src';
import { BrowserTracker } from '@snowplow/browser-tracker-core';
import { setImmediate } from 'timers';

declare var jsdom: JSDOM;

describe('Privacy Sandbox plugin', () => {
  it('Returns values for Privacy Sandbox properties', async () => {
    const sampleTopicEntry = {
      topic: 1,
      taxonomyVersion: '1',
      version: 'chrome.1:1:1',
      modelVersion: '1',
      configurationVersion: '1',
    };

    Object.defineProperties(jsdom.window.document, {
      browsingTopics: {
        value: () => {
          return Promise.resolve([sampleTopicEntry]);
        },
        configurable: true,
      },
      featurePolicy: {
        value: {
          allowsFeature: (feature: string) => feature === 'browsing-topics',
        },
        configurable: true,
      },
    });

    const plugin = PrivacySandboxPlugin();

    const core = trackerCore({
      corePlugins: [plugin],
      callback: (payloadBuilder) => {
        const json = payloadBuilder.getJson().filter((e) => e.keyIfEncoded === 'cx');
        expect(json[0].json).toMatchSnapshot();
      },
    });

    plugin.activateBrowserPlugin?.({ core } as BrowserTracker);

    const flushPromises = () => new Promise(setImmediate);
    await flushPromises();

    core.track(buildLinkClick({ targetUrl: 'https://example.com' }));
  });

  it('Returns no values when feature policy "browsing-topics" is off', async () => {
    const sampleTopicEntry = {
      topic: 1,
      taxonomyVersion: '1',
      version: 'chrome.1:1:1',
      modelVersion: '1',
      configurationVersion: '1',
    };

    Object.defineProperties(jsdom.window.document, {
      browsingTopics: {
        value: () => {
          return Promise.resolve([sampleTopicEntry]);
        },
        configurable: true,
      },
      featurePolicy: {
        value: {
          allowsFeature: function (feature: string) {
            if (feature === 'browsing-topics') {
              return false;
            }
            return true;
          },
        },
        configurable: true,
      },
    });

    const plugin = PrivacySandboxPlugin();

    const core = trackerCore({
      corePlugins: [plugin],
      callback: (payloadBuilder) => {
        const json = payloadBuilder.getJson().filter((e) => e.keyIfEncoded === 'cx');
        expect(json[0]?.json).toMatchSnapshot();
      },
    });

    plugin.activateBrowserPlugin?.({ core } as BrowserTracker);

    const flushPromises = () => new Promise(setImmediate);
    await flushPromises();

    core.track(buildLinkClick({ targetUrl: 'https://example.com' }));
  });
});
