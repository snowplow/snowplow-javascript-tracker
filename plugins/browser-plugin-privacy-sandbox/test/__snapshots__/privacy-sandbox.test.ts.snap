// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Privacy Sandbox plugin Returns no values when feature policy "browsing-topics" is off 1`] = `undefined`;

exports[`Privacy Sandbox plugin Returns values for Privacy Sandbox properties 1`] = `
Object {
  "data": Array [
    Object {
      "data": Object {
        "topics": Array [
          Object {
            "topic": 1,
            "version": "chrome.1:1:1",
          },
        ],
      },
      "schema": "iglu:com.google.privacysandbox/topics/jsonschema/1-0-0",
    },
  ],
  "schema": "iglu:com.snowplowanalytics.snowplow/contexts/jsonschema/1-0-0",
}
`;
